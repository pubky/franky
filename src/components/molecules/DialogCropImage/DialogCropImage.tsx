'use client';

import { useCallback, useEffect, useMemo, useState } from 'react';
import Cropper, { type Area } from 'react-easy-crop';
// NOTE: CSS import is localized here since this is currently the only component using the cropper.
// If multiple components start using react-easy-crop, consider moving this CSS import to:
// - app/globals.css (for global availability)
// - A shared client layout/provider (for centralized client-side loading)
// This prevents duplicate style inclusion and ensures consistent styling across the app.
import 'react-easy-crop/react-easy-crop.css';

import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

type DialogCropImageProps = {
  open: boolean;
  imageSrc: string | null;
  fileName: string;
  fileType: string;
  onClose: () => void;
  onBack: () => void;
  onCrop: (file: File, previewUrl: string) => void;
};

const DEFAULT_FILE_NAME = 'cropped-image.png';

export function DialogCropImage({ open, imageSrc, fileName, fileType, onClose, onBack, onCrop }: DialogCropImageProps) {
  const [crop, setCrop] = useState({ x: 0, y: 0 });
  const [zoom, setZoom] = useState(1);
  const [croppedAreaPixels, setCroppedAreaPixels] = useState<Area | null>(null);
  const [isCropping, setIsCropping] = useState(false);

  useEffect(() => {
    if (open) {
      setCrop({ x: 0, y: 0 });
      setZoom(1);
      setCroppedAreaPixels(null);
      setIsCropping(false);
    }
  }, [open, imageSrc]);

  const handleCropComplete = useCallback((_croppedArea: Area, croppedAreaPixelsValue: Area) => {
    setCroppedAreaPixels(croppedAreaPixelsValue);
  }, []);

  const hasImage = useMemo(() => Boolean(imageSrc), [imageSrc]);

  const handleDone = useCallback(async () => {
    if (!imageSrc || !croppedAreaPixels) return;

    try {
      setIsCropping(true);
      const targetMimeType = fileType && fileType.startsWith('image/') ? fileType : 'image/png';
      const blob = await Libs.cropImageToBlob(imageSrc, croppedAreaPixels, targetMimeType);
      const resolvedMimeType = blob.type || targetMimeType;
      const croppedFile = new File([blob], fileName || DEFAULT_FILE_NAME, { type: resolvedMimeType });
      const previewUrl = URL.createObjectURL(blob);
      onCrop(croppedFile, previewUrl);
    } catch (error) {
      Libs.Logger.error('Failed to crop image', error);
    } finally {
      setIsCropping(false);
    }
  }, [croppedAreaPixels, fileName, fileType, imageSrc, onCrop]);

  return (
    <Atoms.Dialog
      open={open}
      onOpenChange={(nextOpen) => {
        if (!nextOpen) onClose();
      }}
    >
      <Atoms.DialogContent
        className="bg-popover border-border p-8 sm:p-6 max-w-xl gap-6 rounded-2xl"
        showCloseButton={false}
      >
        <Atoms.DialogHeader className="gap-1">
          <Atoms.DialogTitle className="text-2xl sm:text-xl">Cropped Image</Atoms.DialogTitle>
          <Atoms.DialogDescription>Adjust the selection to create the perfect square avatar.</Atoms.DialogDescription>
        </Atoms.DialogHeader>

        <div className="flex flex-col gap-4">
          <div className="relative mx-auto aspect-square w-full max-w-sm overflow-hidden rounded-2xl bg-black/60">
            {hasImage ? (
              <Cropper
                image={imageSrc ?? undefined}
                crop={crop}
                zoom={zoom}
                aspect={1}
                onCropChange={setCrop}
                onZoomChange={setZoom}
                onCropComplete={handleCropComplete}
                restrictPosition
                objectFit="contain"
                showGrid
              />
            ) : (
              <div className="flex h-full w-full items-center justify-center text-sm text-muted-foreground">
                Select an image to start cropping
              </div>
            )}

            <Atoms.Button
              type="button"
              variant="ghost"
              size="sm"
              className="absolute bottom-4 left-4 rounded-full bg-black/50 px-4 text-white hover:bg-black/70"
              onClick={onBack}
            >
              ← Back
            </Atoms.Button>
          </div>

          <div className="flex items-center gap-4">
            <span id="zoom-label" className="text-xs font-medium uppercase tracking-wide text-muted-foreground">
              Zoom
            </span>
            <input
              type="range"
              aria-labelledby="zoom-label"
              min={1}
              max={3}
              step={0.01}
              value={zoom}
              onChange={(event) => setZoom(Number(event.target.value))}
              className="w-full accent-brand"
            />
          </div>
        </div>

        <Atoms.DialogFooter className="gap-4">
          <Atoms.Button type="button" variant="outline" size="lg" className="flex-1 rounded-full" onClick={onClose}>
            Cancel
          </Atoms.Button>
          <Atoms.Button
            type="button"
            size="lg"
            className="flex-1 rounded-full"
            onClick={handleDone}
            disabled={!hasImage || !croppedAreaPixels || isCropping}
          >
            {isCropping ? 'Processing…' : 'Done'}
          </Atoms.Button>
        </Atoms.DialogFooter>
      </Atoms.DialogContent>
    </Atoms.Dialog>
  );
}
