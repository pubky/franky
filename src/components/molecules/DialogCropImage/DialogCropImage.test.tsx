import { describe, it, expect, vi, afterEach } from 'vitest';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import type { Area } from 'react-easy-crop';
import { DialogCropImage } from './DialogCropImage';

const mockCroppedPixels: Area = { width: 120, height: 120, x: 10, y: 15 };

vi.mock('react-easy-crop/react-easy-crop.css', () => ({}));

let latestCropComplete: ((area: Area, croppedAreaPixels: Area) => void) | null = null;

vi.mock(
  'react-easy-crop',
  () => ({
    __esModule: true,
    default: ({ onCropComplete }: { onCropComplete: (area: Area, croppedAreaPixels: Area) => void }) => {
      latestCropComplete = onCropComplete;
      return <div data-testid="cropper" />;
    },
  }),
  { virtual: true },
);

const triggerCropComplete = () => {
  if (latestCropComplete) {
    act(() => {
      latestCropComplete?.(mockCroppedPixels, mockCroppedPixels);
    });
  }
};

const mockCropImageToBlob = vi.fn();

vi.mock('@/libs', () => ({
  cropImageToBlob: (...args: unknown[]) => mockCropImageToBlob(...args),
}));

vi.mock('@/atoms', () => ({
  Dialog: ({ children }: { children: React.ReactNode }) => <div data-testid="dialog">{children}</div>,
  DialogContent: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="dialog-content" className={className}>
      {children}
    </div>
  ),
  DialogHeader: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="dialog-header" className={className}>
      {children}
    </div>
  ),
  DialogTitle: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <h2 data-testid="dialog-title" className={className}>
      {children}
    </h2>
  ),
  DialogDescription: ({ children }: { children: React.ReactNode }) => (
    <p data-testid="dialog-description">{children}</p>
  ),
  DialogFooter: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="dialog-footer" className={className}>
      {children}
    </div>
  ),
  Button: ({
    children,
    onClick,
    disabled,
    className,
  }: {
    children: React.ReactNode;
    onClick?: () => void;
    disabled?: boolean;
    className?: string;
  }) => (
    <button type="button" data-testid="button" onClick={onClick} disabled={disabled} className={className}>
      {children}
    </button>
  ),
}));

if (!('createObjectURL' in URL)) {
  Object.defineProperty(URL, 'createObjectURL', {
    configurable: true,
    writable: true,
    value: vi.fn(),
  });
}

const createDefaultProps = () => ({
  open: true,
  imageSrc: 'data:image/png;base64,example',
  fileName: 'avatar.png',
  fileType: 'image/png',
  onClose: vi.fn(),
  onBack: vi.fn(),
  onCrop: vi.fn(),
});

afterEach(() => {
  latestCropComplete = null;
  vi.clearAllMocks();
});

describe('DialogCropImage', () => {
  it('matches snapshot with image loaded', async () => {
    const props = createDefaultProps();
    const { container } = render(<DialogCropImage {...props} />);

    triggerCropComplete();

    await waitFor(() => {
      const doneButton = screen.getByRole('button', { name: 'Done' });
      expect(doneButton).not.toBeDisabled();
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('invokes crop callback when Done is clicked', async () => {
    const props = createDefaultProps();
    const blob = new Blob(['test'], { type: 'image/jpeg' });
    mockCropImageToBlob.mockResolvedValueOnce(blob);
    const createObjectURLSpy = vi.spyOn(URL, 'createObjectURL').mockReturnValue('blob:preview-url');

    render(<DialogCropImage {...props} />);

    triggerCropComplete();

    const doneButton = await screen.findByRole('button', { name: 'Done' });

    fireEvent.click(doneButton);

    await waitFor(() => {
      expect(mockCropImageToBlob).toHaveBeenCalledWith(props.imageSrc, mockCroppedPixels, props.fileType);
      expect(props.onCrop).toHaveBeenCalled();
    });

    const [fileArg, previewArg] = props.onCrop.mock.calls[0];
    expect(fileArg).toBeInstanceOf(File);
    expect(fileArg.type).toBe('image/jpeg');
    expect(fileArg.name).toBe(props.fileName);
    expect(previewArg).toBe('blob:preview-url');

    createObjectURLSpy.mockRestore();
  });

  it('triggers onBack when back button is pressed', async () => {
    const props = createDefaultProps();
    render(<DialogCropImage {...props} />);

    triggerCropComplete();

    const backButton = await screen.findByRole('button', { name: '‚Üê Back' });
    fireEvent.click(backButton);

    expect(props.onBack).toHaveBeenCalledTimes(1);
  });
});
