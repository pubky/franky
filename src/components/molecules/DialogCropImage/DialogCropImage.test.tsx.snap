// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DialogCropImage > matches snapshot with image loaded 1`] = `
<div
  data-testid="dialog"
>
  <div
    class="bg-popover border-border p-8 sm:p-6 max-w-xl gap-6 rounded-2xl"
    data-testid="dialog-content"
  >
    <div
      class="gap-1"
      data-testid="dialog-header"
    >
      <h2
        class="text-2xl sm:text-xl"
        data-testid="dialog-title"
      >
        Cropped Image
      </h2>
      <p
        data-testid="dialog-description"
      >
        Adjust the selection to create the perfect square avatar.
      </p>
    </div>
    <div
      class="flex flex-col gap-4"
    >
      <div
        class="relative mx-auto aspect-square w-full max-w-sm overflow-hidden rounded-2xl bg-black/60"
      >
        <div
          data-testid="cropper"
        />
        <button
          class="absolute bottom-4 left-4 rounded-full bg-black/50 px-4 text-white hover:bg-black/70"
          data-testid="button"
          type="button"
        >
          ‚Üê Back
        </button>
      </div>
      <div
        class="flex items-center gap-4"
      >
        <span
          class="text-xs font-medium uppercase tracking-wide text-muted-foreground"
          id="zoom-label"
        >
          Zoom
        </span>
        <input
          aria-labelledby="zoom-label"
          class="w-full accent-brand"
          max="3"
          min="1"
          step="0.01"
          type="range"
          value="1"
        />
      </div>
    </div>
    <div
      class="gap-4"
      data-testid="dialog-footer"
    >
      <button
        class="flex-1 rounded-full"
        data-testid="button"
        type="button"
      >
        Cancel
      </button>
      <button
        class="flex-1 rounded-full"
        data-testid="button"
        type="button"
      >
        Done
      </button>
    </div>
  </div>
</div>
`;
